### DISCUSSION TITLE
conversion of cudipy.align.imwarp.DiffeomorphicMap to dipy.align.imwarp.DiffeomorphicMap

### DISCUSSION THREAD

**Discussion by rajnraj100** [dipy/dipy#3137] [Registration]
Hi...Can we transform  cudipy.align.imwarp.DiffeomorphicMap to  dipy.align.imwarp.DiffeomorphicMap?

**Comment by skoudoro** [comment]
Hi @rajnraj100,

it seems to be exactly the same so yes you can transform `cudipy.align.imwarp.DiffeomorphicMap` to `dipy.align.imwarp.DiffeomorphicMap`

**Reply by rajnraj100** [reply]
If I use Diffeomorphic registration part in CPU using dipy,  I would get:
>>from dipy.align.metrics import CCMetric
>>from dipy.align.imwarp import SymmetricDiffeomorphicRegistration
>>metric = CCMetric(3)
>>level_iters = [10, 10, 5]
>>pre_align = 4x4 numpy array (numpy.ndarray)
>>sdr = SymmetricDiffeomorphicRegistration(metric, level_iters)
>>cpu_mapping = sdr.optimize(static, moving, static_affine, moving_affine, pre_align)
>>type(cpu_mapping )
dipy.align.imwarp.DiffeomorphicMap

But, I have done my Diffeomorphic registration part in GPU using cudipy for speeding up my code. So,  I am getting:
>>from cudipy.align.metrics import CCMetric
>>from cudipy.align.imwarp import SymmetricDiffeomorphicRegistration
>>metric = CCMetric(3)
>>level_iters = [10000, 1000, 100]
>>sdr = SymmetricDiffeomorphicRegistration(metric, level_iters)
>>gpu_mapping = sdr.optimize( static_arr, moving_arr, static_affine, moving_affine, pre_align  )
>>type(gpu_mapping )
cudipy.align.imwarp.DiffeomorphicMap

Here, I stored mapping for all my subjects' data. I will apply mapping.transorm(.) or mapping.transform_inverse(.) later as per my need.
Now, for some reason, I need to switch to cpu for rest of my code but for that I need to convert this cudipy.align.imwarp.DiffeomorphicMap to dipy.align.imwarp.DiffeomorphicMap so that I will not have to use gpu anymore. 
Here, I am stuck as I do not know how to change it from cudipy (GPU) to dipy (CPU):

cpu_mapping = some_cmd(gpu_mapping ) 

If it is possible , kindly help me with the command/ code. Thank you.

**Reply by pjsjongsung** [reply]
You could transfer all the parameters of the DiffeomorphicMap that are located in the gpu(cudipy class) and copy them to the cpu(dipy class). I know it is not trivial, but if you can copy all the arrays that were calculated through optimize, it should be serve as a workaround for now.

**Reply by rajnraj100** [reply]
This will not serve the purpose. Anyways, thank you for responding.

**Reply by skoudoro** [reply]
> this will not serve the purpose. Anyways, thank you for responding.

not sure why? Can you explain why it will not serve the purpose.  I believe you just need to initialize the `DiffeomorphicMap` and then update the `DiffeomorphicMap.forward` and `DiffeomorphicMap.backward` field.  I believe nothing more is needed between both implementation. 

when applying the transformation `backward`or `forward` will be used.

**Comment by arokem** [comment]
I'm not so sure that I understand what the word "transform" means here. Do you mean whether you could use them interchangeably? I think the answer to that is yes. If the question is whether we should replace our current implementation in DIPY with the one implemented in cuDIPY, I think that I would be against that, at least for the time being. I think that we want to keep the GPU-accelerated functionality separate for now, because of the complexities of installing and running these implementations.

**Reply by rajnraj100** [reply]
If I use Diffeomorphic registration part in CPU using dipy,  I would get:
>>from dipy.align.metrics import CCMetric
>>from dipy.align.imwarp import SymmetricDiffeomorphicRegistration
>>metric = CCMetric(3)
>>level_iters = [10, 10, 5]
>>pre_align = 4x4 numpy array (numpy.ndarray)
>>sdr = SymmetricDiffeomorphicRegistration(metric, level_iters)
>>cpu_mapping = sdr.optimize(static, moving, static_affine, moving_affine, pre_align)
>>type(cpu_mapping )
dipy.align.imwarp.DiffeomorphicMap

But, I have done my Diffeomorphic registration part in GPU using cudipy for speeding up my code. So,  I am getting:
>>from cudipy.align.metrics import CCMetric
>>from cudipy.align.imwarp import SymmetricDiffeomorphicRegistration
>>metric = CCMetric(3)
>>level_iters = [10000, 1000, 100]
>>sdr = SymmetricDiffeomorphicRegistration(metric, level_iters)
>>gpu_mapping = sdr.optimize( static_arr, moving_arr, static_affine, moving_affine, pre_align  )
>>type(gpu_mapping )
cudipy.align.imwarp.DiffeomorphicMap

Here, I stored mapping for all my subjects' data. I will apply mapping.transorm(.) or mapping.transform_inverse(.) later as per my need.
Now, for some reason, I need to switch to cpu for rest of my code but for that I need to convert this cudipy.align.imwarp.DiffeomorphicMap to dipy.align.imwarp.DiffeomorphicMap so that I will not have to use gpu anymore. 
Here, I am stuck as I do not know how to change it from cudipy (GPU) to dipy (CPU):

cpu_mapping = some_cmd(gpu_mapping )

If it is possible , kindly help me with the command/ code. Thank you.
